### PR TITLE
Feature/LevelLoader into develop

### PR DESCRIPTION
**Pull request by skoolaidl**
Loading in of all level components via XML. Utilizes TinyXML2 library.

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -1 +1,2 @@
 Debug/
+.DS_Store
```
Comment (jkula25) [inline .gitignore]: I'm pretty sure you should delete this file from the pull request

File: include/GameLogic.h [hunk 2]
```diff
@@ -27,16 +28,14 @@ class GameLogic {
         Enemy enemy4;
         Platform floor;
         Platform platformA;
```
Comment (jkula25) [inline include/GameLogic.h]: Why are only some of these objects deleted?

File: include/GameLogic.h [hunk 3]
```diff
@@ -27,16 +28,14 @@ class GameLogic {
...
-        Projectile projectile;
-        Spike spike1;
-        Spike spike2;
-        Spike spike3;
-        Spike spike4;
+
+        LevelCreator loader;
+
         std::vector<std::reference_wrapper<Actor>> actorsVector;
+        std::vector<std::vector<Actor>> nestedActors;
```
Comment (jkula25) [inline include/GameLogic.h]: What is this vector of vectors used for?

File: include/GameLogic.h [hunk 4]
```diff
@@ -27,16 +28,14 @@ class GameLogic {
...
+        std::vector<Platform> platVector;
+        std::vector<Spike> spikeVector;
+        std::vector<Enemy> enemyVector;
```
Comment (jkula25) [inline include/GameLogic.h:37]: What is the purpose of all these seemingly duplicate vectors? Also, rename platVector to platformVector.

File: include/enemy.h [hunk 5]
```diff
@@ -12,6 +12,7 @@ class Enemy : public Moveable {
         float velocityY;
         float xpos;
         float ypos;
+        int colorType;
```
Comment (jkula25) [inline include/enemy.h]: I have a variable just like this in my branch but it's just called type. So I would rename this to "type" to avoid merge conflicts.

File: include/platform.h [hunk 6]
```diff
@@ -1,6 +1,7 @@
 #ifndef PLATFORM_H
 #define PLATFORM_H
 #include <SFML/Graphics.hpp>
+#include <SFML/System.hpp>
```
Comment (jkula25) [inline include/platform.h]: Why is this needed?

File: res/MapTest.txt [hunk 7]
```diff
@@ -0,0 +1,11 @@
+x,x x,x x,x x,x x,x x,x x,x x,x x,x x,x x,x x,x x,x x,x x,x x,x
```
Comment (jkula25) [inline res/MapTest.txt]: Why is this file included if we switched to using tinyxml2?

File: src/GameLogic.cpp [hunk 8]
```diff
@@ -102,6 +56,28 @@ void GameLogic::update(float timeS) {
 void GameLogic::setLevel(int level) {
     currentLevel = level;
     //load specified level
+    loader.LoadLevel(level);
+    platVector = loader.getPlatforms();
+    spikeVector = loader.getSpikes();
+    enemyVector = loader.getEnemies();
```
Comment (jkula25) [inline src/GameLogic.cpp:62]: I would reduce the scope of these variables to being declared just in this function since they are not being used anywhere else in GameLogic and do not need to be member variables

File: src/GameLogic.cpp [hunk 9]
```diff
@@ -182,7 +158,7 @@ void GameLogic::updateProjectileCollisions() {
 
 void GameLogic::removeFromActorsVector(Actor& actor) {
     for (int i = 0; i < actorsVector.size(); ++i) {
-        if (actor == actorsVector[i].get()) {
+        if (actor == actorsVector[i]) {
```
Comment (jkula25) [inline src/GameLogic.cpp]: Why is this changed? Because actorsVector[i] would point to the reference wrapper and not the actual actor

File: src/LevelCreator.cpp [hunk 10]
```diff
@@ -0,0 +1,114 @@
+#include "LevelCreator.h"
+
+
+/* 
+Level Creator class serves as an engine for reading in a specific level text file
+and loading the proper tiles for that level to display to the player.
```
Comment (jkula25) [inline src/LevelCreator.cpp]: I'd update this comment according to using XML for creating levels

File: src/LevelCreator.cpp [hunk 11]
```diff
@@ -0,0 +1,114 @@
...
+
+void LevelCreator::LoadLevel(int level){
+    //This line will change based on "level", for now just using MapTest
+    // std::ifstream openfile("../res/MapTest.txt");
+
+    //This line will change based on "level", for now just using Sample Level XML
+    tinyxml2::XMLDocument map;
+    tinyxml2::XMLError eResult = map.LoadFile("../res/practice_level.xml");
+    if (eResult != tinyxml2::XML_SUCCESS) {
+        std::cout << "error loading file" << std::endl;
```
Comment (jkula25) [inline src/LevelCreator.cpp:21]: I would put a return statement here so that way the function exits when it fails

File: src/LevelCreator.cpp [hunk 12]
```diff
@@ -0,0 +1,114 @@
...
+
+
+    if (entry) {
+        for (tinyxml2::XMLNode *node = entry->FirstChildElement(); node; node = node->NextSibling()){
+
+            tinyxml2::XMLElement *e = node->ToElement();
+
+            std::cout << e->Value() << std::endl;
+
+            if (std::strncmp(e->Value(), "Platform", 9) == 0) {
```
Comment (jkula25) [inline src/LevelCreator.cpp]: If you are doing a simple string compare, just use "==" or at the very least I believe you can use strcmp since you are comparing the whole word and not a specific amount of characters

File: src/LevelCreator.cpp [hunk 13]
```diff
@@ -0,0 +1,114 @@
...
+                float y = std::stof(stringLocY);
+                
+
+                Platform platform;
+                platform.init(scaleX, scaleY, x, y);
+                platforms.push_back(platform);
+
+            }
+
+            else if (std::strncmp(e->Value(), "Spike", 6) == 0) {
```
Comment (jkula25) [inline src/LevelCreator.cpp]: Same as previous comment

File: src/LevelCreator.cpp [hunk 14]
```diff
@@ -0,0 +1,114 @@
...
+                float x = std::stof(stringLocX);
+                float y = std::stof(stringLocY);
+
+                Spike spike;
+                spike.init(scaleX, scaleY, x, y);
+                spikes.push_back(spike);
+
+            }
+
+            else if (std::strncmp(e->Value(), "Enemy", 6) == 0) {
```
Comment (jkula25) [inline src/LevelCreator.cpp]: Again same as previous comment

File: src/LevelCreator.cpp [hunk 15]
```diff
@@ -0,0 +1,114 @@
...
+
+                float x = std::stof(stringLocX);
+                float y = std::stof(stringLocY);
+                int color = std::stoi(stringColor);
+
+                Enemy enemy;
+                enemy.init(x, y, color);
+                enemies.push_back(enemy);
+            }
+            else if (std::strncmp(e->Value(), "Girl", 4) == 0) {
```
Comment (jkula25) [inline src/LevelCreator.cpp]: Again same as previous comment

File: src/enemy.cpp [hunk 16]
```diff
@@ -33,6 +33,7 @@ void Enemy::init() {
 }
 
 void Enemy::init(float x, float y, int color) {
+    colorType = color;
```
Comment (jkula25) [inline src/enemy.cpp]: Again, I would change this to just "type" for uniformity

File: include/GameLogic.h (hunk 4 above)
Comment (skoolaidl) [reply include/GameLogic.h:37]: Yeah it looks like they are duplicates of the following vectors, however these are different because they don't use reference wrappers. @carter-greis knows better than I do about the difference in functionality, but I know that when loading in the object from level loader they cannot go directly into the reference wrapper.

edit: Fixed the naming of platVector though

File: include/platform.h (hunk 6 above)
Comment (skoolaidl) [reply include/platform.h]: Not sure, I didn't add these...it's in a couple of header files though

Comment (skoolaidl) [reply include/platform.h]: @carter-greis do you know?

File: include/GameLogic.h (hunk 4 above)
Comment (carter-greis) [reply include/GameLogic.h:37]: So I tried defining them just locally in the setLevel method, but then the objects weren't staying defined after the method ended and it wasn't working. I still think the conversion to reference_wrapper is needed because otherwise the objects stored in the actorsVector and the individual vectors for the specific types would not be the same object and thus changes made from the specific vectors would not affect the appropriate object in the actorsVector. I believe that is why I originally made them reference_wrappers. I am travelling today and won't have access to my computer until tonight, but I will try to take a look at it and make sure that my reasoning is still valid.

File: src/LevelCreator.cpp (hunk 12 above)
Comment (skoolaidl) [reply src/LevelCreator.cpp]: I believe in C and C++ you cannot use the "==" for string comparisons. I can change it to strcmp though

File: src/GameLogic.cpp (hunk 9 above)
Comment (skoolaidl) [reply src/GameLogic.cpp]: Not sure, both seem to work. What do you think?

File: src/LevelCreator.cpp (hunk 11 above)
Comment (skoolaidl) [reply src/LevelCreator.cpp:21]: Done

File: src/GameLogic.cpp (hunk 8 above)
Comment (skoolaidl) [reply src/GameLogic.cpp:62]: Doing this is causing a seg fault, and when I declare them locally with a specific size I am getting a Bus Error 10. I don't think I understand enough about C++ to know why this might be happening, any suggestions?

File: include/platform.h (hunk 6 above)
Comment (jkula25) [reply include/platform.h]: I would just remove it and if it compiles fine, then it's not needed. I'm pretty sure it's for handling window properties

File: src/GameLogic.cpp (hunk 9 above)
Comment (jkula25) [reply src/GameLogic.cpp]: I would have it be actorsVector[i].get() because that points to an actor versus without the .get() points to a reference wrapper

File: src/LevelCreator.cpp (hunk 12 above)
Comment (jkula25) [reply src/LevelCreator.cpp]: In Java you can't use "==" since that operator compares the memory address instead of the string's contents, but in C++ "==" essentially just calls the compare function and checks that it equals 0. So "==" in C++ does compare the string's contents, so you should use "==" since it's more readable

File: src/GameLogic.cpp (hunk 9 above)
Comment (carter-greis) [reply src/GameLogic.cpp]: I think either way works because c++ automatically converts the reference_wrapper to an Actor implicitly in the background, but using .get() will make more sense, as the actor is explicitly what we want the == to compare.

File: src/GameLogic.cpp (hunk 8 above)
Comment (carter-greis) [reply src/GameLogic.cpp:62]: They cannot be declared locally because then they go out of scope as soon as that local scope ends and they get destroyed. This is the easiest way around that without having to start allocating memory space on the heap which opens up a whole host of problems with memory management and memory leaks. It is also the reason why I defined them like this. I am pretty sure that we still need the vectors of reference_wrappers though. I am going to go ahead and empty them after we use them in setLevel though, so that they are not taking up space in memory.

Comment (carter-greis) [reply src/GameLogic.cpp:62]: Actually, now that I think about it, I think I can change it so that platforms, spikes, and enemies are normal vectors and only actorsVector is a vector of reference_wrappers, since actorsVector is the only one that gets passed out of GameLogic to HumanView. I will go ahead and work on that.

File: src/LevelCreator.cpp (hunk 12 above)
Comment (carter-greis) [reply src/LevelCreator.cpp]: That doesn't work here though because the Value() method for tinyxml returns a char* that represents an array of chars and c++ treats literal strings as char arrays as well. Thus, it does an == between two char arrays and the == method for string never gets called because neither of them are actually strings. std::strcmp accepts strings and char arrays for input, which is why it works.

File: src/GameLogic.cpp (hunk 9 above)
Comment (skoolaidl) [reply src/GameLogic.cpp]: Done
